src/Microsoft.TemplateEngine.Edge/Settings/GlobalSettings.cs

```diff
@@ -0,0 +1,150 @@
+            _locked = true;
+            return mutex;
+        }
+
+        private void Unlocked()
+        {
+            _locked = false;
+        }
+
+        public void Dispose()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+            _disposed = true;
+
+            _watcher?.Dispose();
+            _watcher = null;
+        }
+
+        public async Task<IReadOnlyList<TemplatePackageData>> GetInstalledTemplatesPackagesAsync(CancellationToken cancellationToken)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(GlobalSettings));
+            }
+
+            if (!_environmentSettings.Host.FileSystem.FileExists(_globalSettingsFile))
+            {
+                return Array.Empty<TemplatePackageData>();
+            }
+
+            for (int i = 0; i < 5; i++)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    string? textFileContent = _paths.ReadAllText(_globalSettingsFile, "{}");
+                    GlobalSettingsData? data = JsonConvert.DeserializeObject<GlobalSettingsData>(textFileContent);
+                    return data.Packages ?? Array.Empty<TemplatePackageData>();
+                }
+                catch (Exception)
+                {
+                    if (i == 4)
+                    {
+                        throw;
+                    }
+                }
+                await Task.Delay(20).ConfigureAwait(false);
+            }
+            throw new InvalidOperationException();
+        }
+
+        public async Task SetInstalledTemplatesPackagesAsync(IReadOnlyList<TemplatePackageData> packages, CancellationToken cancellationToken)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(GlobalSettings));
+            }
+
+            if (!_locked)
+            {
+                throw new InvalidOperationException($"Before calling {nameof(SetInstalledTemplatesPackagesAsync)}, {nameof(LockAsync)} must be called.");
+            }
+
+            string? serializedText = JsonConvert.SerializeObject(new GlobalSettingsData()
+            {
+                Packages = packages
+            });
+
+            for (int i = 0; i < 5; i++)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    _paths.WriteAllText(_globalSettingsFile, serializedText);
+                    SettingsChanged?.Invoke();
+                    return;
+                }
+                catch (Exception)
+                {
+                    if (i == 4)
+                    {
+                        throw;
+                    }
+                }
+                await Task.Delay(20).ConfigureAwait(false);
```

Comment:
[Random foray into the implementation]
Why not pass the `cancellationToken` to `Delay` here?

Reply:
Added, tnx!